tunnel/tunnel_reverse_net.go

```diff
@@ -25,110 +25,125 @@ func (h *ReverseForwardingHandler) HandleSSHRequest(ctx ssh.Context, srv *ssh.Se
-
-			// Implement GetUpstreamConn by opening a channel on the SSH connection.
-			GetUpstreamConn: func(tConn net.Conn) (io.ReadWriteCloser, error) {
-				// Address of tunnel client.
-				originAddr, originPortStr, _ := net.SplitHostPort(tConn.RemoteAddr().String())
-				originPort, _ := strconv.Atoi(originPortStr)
-
-				// Construct port forwarding response
-				payload := remoteForwardChannelData{
-					// Tunnel listener address
-					DestAddr: reqPayload.BindAddr,
-					DestPort: reqPayload.BindPort,
-
-					// Tunnel client address
-					OriginAddr: originAddr,
-					OriginPort: uint32(originPort),
-				}
-
-				// Open SSH channel.
-				ch, reqs, err := conn.OpenChannel(forwardedTCPChannelType, gossh.Marshal(payload))
-				if err != nil {
-					return nil, errors.Wrap(err, "could not open channel")
-				}
-				// Discard all other incoming requests.
-				go gossh.DiscardRequests(reqs)
-
-				return ch, nil
-			},
-
-			Lifecycle: h.lifecycle,
-			Stats:     h.stats,
+		return h.openPortForwarding(ctx, payload)
+
+	// Handle a request from the SSH connection to close port forwarding.
+	case sshTCPForwardCloseEvent:
+		var payload remoteForwardCancelRequest
+		if err := gossh.Unmarshal(req.Payload, &payload); err != nil {
+			return false, []byte{}
 		}
+		return h.closePortForwarding(ctx, payload)
 
-		// Start tunnel listener
-		if err := forwarder.Listen(); err != nil {
-			switch err.(type) {
-			case bootError:
-				h.lifecycle.BootError(err)
-			default:
-				h.lifecycle.Error(err)
-			}
+	default:
+		return false, nil
+	}
+}
 
-			return false, []byte("bind address already in use")
-		}
+// closePortForwarding handles a request from the SSH client to open port forwarding
+func (h *ReverseForwardingHandler) openPortForwarding(ctx context.Context, payload remoteForwardOpenRequest) (bool, []byte) {
+	tunnel, ok := h.GetTunnel(int(payload.BindPort))
+	if !ok {
+		// Couldn't find a valid registered tunnel for this request
+		return false, []byte("Port forwarding is disabled")
+	}
 
-		h.Lock()
-		h.forwards[bindAddr] = forwarder
-		h.Unlock()
-
-		// Start port forwarding
-		go forwarder.Serve()
-
-		// Graceful shutdown if connection ends
-		go func() {
-			<-ctx.Done()
-			h.Lock()
-			f, ok := h.forwards[bindAddr]
-			h.Unlock()
-			if ok {
-				f.Close()
+	// Get the connection out of the context
+	conn := ctx.Value(ssh.ContextKeyConn).(*gossh.ServerConn)
+	tunnelBindAddr := net.JoinHostPort(payload.BindAddr, strconv.Itoa(int(payload.BindPort)))
+
+	// Initiate TCPForwarder to listen for tunnel connections.
+	forwarder := &TCPForwarder{
+		BindAddr:         tunnelBindAddr,
+		HTTPProxyEnabled: h.httpProxyEnabled,
+
+		// Implement GetUpstreamConn by opening a channel on the SSH connection.
+		GetUpstreamConn: func(tConn net.Conn) (io.ReadWriteCloser, error) {
+			// Address of tunnel client.
+			originAddr, originPortStr, _ := net.SplitHostPort(tConn.RemoteAddr().String())
+			originPort, _ := strconv.Atoi(originPortStr)
```

Comment:
swallowing the error here, are we safe because of how the `originPortStr` is generated is from `tConn.RemoteAddr().String()`?